src/compute/src/sink/persist_sink.rs

```diff
@@ -1105,12 +1125,66 @@ where
                         }
                     }
                 }
+                _it_changed = read_only.changed() => {
+                    // We might have to append some batches that we stashed
+                    // while in read-only mode.
+                }
                 else => {
                     // All inputs are exhausted, so we can shut down.
                     return;
                 }
             };
 
+            // Only retain descriptions and batches that still have a chance of
+            // being applied.
+            let persist_upper = write.shared_upper();
+            in_flight_descriptions.retain(|(_lower, upper)| {
+                PartialOrder::less_equal(&persist_upper, upper)
+            });
```

Comment:
Same here, I think we should compare with the lower rather than the upper of the batch description.